enhancements/maintenance/machine-maintenance-operator.md

```diff
@@ -0,0 +1,241 @@
+---
+title: machine-maintenance-operator
+authors:
+  - "@dofinn"
+reviewers:
+  - "@cblecker"
+  - "@jharrington22"
+  - "@jewzaam"
+  - "@jeremyeder"
+  - "@michaelgugino"
+  - "@derekwaynecarr"
+approvers:
+  - "@michaelgugino"
+  - "@derekwaynecarr"
+  - "@bison"
+  - "@enxebre"
+creation-date: 2020-05-26
+last-updated: 2020-05-26
+status: provisional
+
+# machine-maintenance-operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+>1. How applicable is this use-case to baremetal installations? 
+
+>2. The ability to declare maintenance windows is currently being discussed in OSD. Nothing has be design in concrete meaning that this proposal has an undesigned depedency. However develop could commence on this operator and have the maintenance functionality retrospectively fitted. Is this a blocker for its acceptenance?
+
+>3. How applicable will this be to a cloud provider like GCP that performs live migrations of their VMs?
+
+## Summary
+
+This enhancement proposal explores the idea of having a machine-maintenance-operator(MMO) that acts as a machine 'nanny'. The MMO will actively search for scheduled maintenances for each infra/worker machine in the cluster along with validating the state of these machines against the cloud providers API. Upon finding a maintenance for a machine, the MMO will seek to execute this maintenance manually at the earliest convenient time as defined by an administrator. 
+
+A POC of the idea can be found [here](https://github.com/dofinn/machine-maintenance-operator).
+
+## Motivation
+
+Currently the machine-api fails to accurately detect the state of machines (validated in AWS) when they have been terminated/stopped by the cloud provider. A provider may terminate/stop and instance when there is an associated maintenance scheduled for it. This is also the case for users manually terminating/stopping machines via the console (validated in AWS). The MMO is a proactive approach to first detecting maintenances and/or machine state from the cloud provider, then executing the required to enable the machine-api to continue managing machines and machinesets. This operator would work in conjunction with MachineHealthCheck implimentation. 
```

Comment:
This is untrue.  We track the instance state every 15 minutes, and it should be available in the status field.  This is often confused with machine phase, which may be "Running" even if the instance is stopped.  As mentioned elsewhere, the word "Running" is not a great fit, what it really means is "MachineBecameANode" and then that phase is unchanged until the machine is deleted.

Reply:
I stand corrected. 

```
oc get machine gxs-aws-nj7hf-worker-ap-southeast-2a-mmjwc -o json | jq '.status.providerStatus.instanceState'
"running"
```

```
oc get machine gxs-aws-nj7hf-worker-ap-southeast-2a-mmjwc -o json | jq '.status.providerStatus.instanceState'
"stopping"
```

```
oc get machine gxs-aws-nj7hf-worker-ap-southeast-2a-mmjwc -o json | jq '.status.providerStatus.instanceState'
"stopped"
```

Is there a remediation path from this state? or is the most agnostic solution just to log the current status and act elsewhere?

Reply:
By default, the machine-controller does not do anything.  Eventually, the node will go unready.  Using MachineHealthChecks will cause the machine to be deleted (if it's part of a machineset), and you'll get a new machine.